### PR TITLE
MINOR: updates docs for KIP-358

### PR DESCRIPTION
**Pull request by mjsax**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @guozhangwang @bbejeck @vvcephei @nizhikov

**Comment by nizhikov** [comment]
Hello, @mjsax 

How can I view docs locally?
My browser shows a white screen when I open HTML file from sources.

**Comment by guozhangwang** [comment]
@nizhikov you can follow this page to setup apache site locally from the source code: https://cwiki.apache.org/confluence/display/KAFKA/Contributing+Website+Documentation+Changes

File: docs/streams/developer-guide/dsl-api.html [hunk 1]
```diff
@@ -1144,17 +1144,17 @@ <h4><a id="streams_concepts_globalktable" href="#streams_concepts_globalktable">
...
 <span class="n">KTable</span><span class="o">&lt;</span><span class="n">Windowed</span><span class="o">&lt;</span><span class="n">String</span><span class="o">&gt;,</span> <span class="n">Long</span><span class="o">&gt;</span> <span class="n">aggregatedStream</span> <span class="o">=</span> <span class="n">groupedStream</span><span class="o">.</span><span class="na">windowedBy</span><span class="o">(</span>
-    <span class="n">TimeWindows</span><span class="o">.</span><span class="na">of</span><span class="o">(</span><span class="n">TimeUnit</span><span class="o">.</span><span class="na">MINUTES</span><span class="o">.</span><span class="na">toMillis</span><span class="o">(</span><span class="mi">5</span><span class="o">)))</span> <span class="cm">/* time-based window */</span>
+    <span class="n">TimeWindows</span><span class="o">.</span><span class="na">of</span><span class="o">(</span><span class="n">Duration</span><span class="o">.</span><span class="na">ofMinutes</span><span class="o">(</span><span class="mi">5</span><span class="o">)))</span> <span class="cm">/* time-based window */</span>
```
Comment (guozhangwang) [inline docs/streams/developer-guide/dsl-api.html:1152]: nit: this is not introduced in this PR, but for `time-based window` we have the last right bracket after the comment, while for session window below the last right bracket is before the comment. Personally I think the latter is better.

File: docs/streams/developer-guide/dsl-api.html [hunk 2]
```diff
@@ -2804,7 +2804,7 @@ <h4><a id="streams_concepts_globalktable" href="#streams_concepts_globalktable">
                         Old records in the state store are purged after the specified
                         <a class="reference internal" href="../core-concepts.html#streams_concepts_windowing"><span class="std std-ref">window retention period</span></a>.
                         Kafka Streams guarantees to keep a window for at least this specified time; the default value is one day and can be
-                        changed via <code class="docutils literal"><span class="pre">Windows#until()</span></code> and <code class="docutils literal"><span class="pre">SessionWindows#until()</span></code>.</p>
+                        changed via <code class="docutils literal"><span class="pre">Materialized#withRetention()</span></code>.</p>
```
Comment (guozhangwang) [inline docs/streams/developer-guide/dsl-api.html:2807]: Nice catch!

File: docs/streams/developer-guide/processor-api.html [hunk 3]
```diff
@@ -365,11 +365,11 @@ <h2><a class="toc-backref" href="#id10">Accessing Processor Context</a><a class=
                 <code class="docutils literal"><span class="pre">headers</span></code>.</p>
             <p>Here is an example implementation of how to add a new header to the record:</p>
             <div class="highlight-java"><div class="highlight"><pre><span></span><span class="n">public void process(String key, String value) {</span>
-            
-            <span class="c1">// add a header to the elements</span>
-            <span class="n">context()</span><span class="o">.</span><span class="na">headers</span><span class="o">()</span><span class="o">.</span><span class="na">add</span><span class="o">.</span><span class="o">(</span><span class="s">&quot;key&quot;</span><span class="o">,</span> <span class="s">&quot;key&quot;</span>
-            <span class="o">}</span>
-            </pre></div>
+
+    <span class="c1">// add a header to the elements</span>
```
Comment (guozhangwang) [inline docs/streams/developer-guide/processor-api.html:369]: Is this intentional?

File: docs/streams/upgrade-guide.html [hunk 4]
```diff
@@ -91,7 +91,24 @@ <h3><a id="streams_api_changes_210" href="#streams_api_changes_210">Streams API
...
     <p>
-        We added a new serde for UUIDs (<code>Serdes.UUIDSerde</code>) that you can use via <code>Serdes.UUID()</code> (cf. <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-206%3A+Add+support+for+UUID+serialization+and+deserialization">KIP-206</a>).
+        We added a new serde for UUIDs (<code>Serdes.UUIDSerde</code>) that you can use via <code>Serdes.UUID()</code>
+	(cf. <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-206%3A+Add+support+for+UUID+serialization+and+deserialization">KIP-206</a>).
+    </p>
+
+    <p>
+        We updated a list of methods that take <code>long</long> arguments as either timestamp (fix point) or duration (time period)
+        and replaced them with <code>Instant</code> and <code>Duration</code> parameters for improved semantics.
+        Some, existing methods are deprecated and users are encouraged to update their code.
```
Comment (guozhangwang) [inline docs/streams/upgrade-guide.html]: `Some, existing` -> `old methods based on `long` are deprecated ...`.

File: docs/streams/upgrade-guide.html [hunk 5]
```diff
@@ -91,7 +91,24 @@ <h3><a id="streams_api_changes_210" href="#streams_api_changes_210">Streams API
...
+        <br />
+        In particular, aggregation windows (hopping/tumbling/unlimited time windows and session windows) as well as join windows now take <code>Duration</code>
+        arguments to specify window size, hop, and gap parameters.
+        Also, window sizes and retention times are now specified as <code>Duration</code> type in <code>Stores</code> class.
+        The <code>Window</code> class has new methods <code>#startTime()</code> and <code>#endTime</code> that return window start/end timestamp as <code>Instant</code>.
+        For interactive queries, there are new <code>#fetch(...)</code> overloads taking <code>Instant</code> arguments.
+        Additionally, punctuations are now registerd via <code>ProcessorContext#schedule(Duration interval, ...)</code>
+        and <code>KafkaStreams#close(Duration)</code> accepts a single timeout argument.
+	Note, that the new <code>#close</code> method has improved (but slightly different) semantics than the old one.
```
Comment (guozhangwang) [inline docs/streams/upgrade-guide.html]: nit: I'd use a new section to emphasize its importance. And also talk about the actual semantic changes here as well.

**Review by guozhangwang — APPROVED** [review]
Feel free to merge after addressing the comments.

File: docs/streams/upgrade-guide.html (hunk 5 above)
Comment (JimGalasyn) [inline docs/streams/upgrade-guide.html]: "Note that" or "Note: "the new"

File: docs/streams/upgrade-guide.html [hunk 6]
```diff
@@ -91,7 +91,24 @@ <h3><a id="streams_api_changes_210" href="#streams_api_changes_210">Streams API
...
+        For more details see <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-358%3A+Migrate+Streams+API+to+Duration+instead+of+long+ms+times">KIP-358</a>.
```
Comment (JimGalasyn) [inline docs/streams/upgrade-guide.html]: "For more details, see ..."

**Review by JimGalasyn — APPROVED** [review]
LGTM, with a couple of nits.

File: docs/streams/developer-guide/dsl-api.html (hunk 1 above)
Comment (mjsax) [reply docs/streams/developer-guide/dsl-api.html:1152]: Not sure what you mean. Both (L1152 and L1157) seem to be the "same" -- or I am looking at the wrong place.

File: docs/streams/developer-guide/processor-api.html (hunk 3 above)
Comment (mjsax) [reply docs/streams/developer-guide/processor-api.html:369]: Yes. Fixed the indention.

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `2.1`.

**Pull request merged**